### ISSUE TITLE
iOS Performance

### ISSUE THREAD

**Issue by Joncom** [Kode/Kha#988]
Hi @RobDangerous.

As you may recall from chat and Github, I've been having performance issues on iOS.

While discussing in chat where the bottleneck might be, you mentioned, "iPhone 5s is a dual core, guys. 50% activity means the thing is CPU bound."

Later, Antriel said, "50% of dual core = one core full. Which means you are bottlenecked, unless you do multithreading."

Obviously a CPU bottleneck is bad, so I started looking into why the CPU would be that high.

One strange thing I noticed is that my custom bunnymark build would be running at 50% CPU before any bunnies have even been added.

![screen shot 2019-02-13 at 3 37 29 pm](https://user-images.githubusercontent.com/1667179/52698363-ecc5b500-2fa5-11e9-94e4-2117f2f32682.png)

To get a better idea of how Kha should be running before adding extra stuff on top (like bunnymark), I tried the [Empty](https://github.com/Kha-Samples/Empty) build. To my surprise, it was running at 100% CPU out of the box. Not sure if that's expected behavior.

However, I remember @sh-dave saying you should draw to the framebuffer every frame, so I patched Empty as follows:

```diff
diff --git a/Sources/Main.hx b/Sources/Main.hx
index 3976b7f..0012af5 100644
--- a/Sources/Main.hx
+++ b/Sources/Main.hx
@@ -11,7 +11,8 @@ class Main {
        }

        static function render(framebuffer: Framebuffer): Void {
-
+               framebuffer.g2.begin(true, kha.Color.Black);
+               framebuffer.g2.end();
        }
```

That seems to have improved things a bit, as now the build only uses 46% CPU.

![screen shot 2019-02-13 at 3 56 24 pm](https://user-images.githubusercontent.com/1667179/52700820-6ca24e00-2fab-11e9-8846-83e9cd692068.png)

However, that's still nearly 50%, which means there isn't much room to run an app on top of Kha, at least on this device?

**Comment by RobDangerous** [comment]
As always, profiling data please.

**Comment by Joncom** [comment]
I'm a little confused, because when profiling, CPU usage hover around 20%, but when running in Xcode (no profiler, it uses more than double that). **Profiling** and **Running** are both configured as "release" builds.

![screen shot 2019-02-13 at 4 55 22 pm](https://user-images.githubusercontent.com/1667179/52703105-2ef3f400-2fb0-11e9-8564-b3f375eca055.png)

Nevertheless, here is what the profiling data shows:

![screen shot 2019-02-13 at 4 57 22 pm](https://user-images.githubusercontent.com/1667179/52703234-75495300-2fb0-11e9-84bf-68e7c33df43d.png)

![screen shot 2019-02-13 at 4 58 08 pm](https://user-images.githubusercontent.com/1667179/52703274-914cf480-2fb0-11e9-93e6-b8b49c2dac2e.png)

**Comment by RobDangerous** [comment]
Looks like it's just waiting for vsync. Not very clever by Xcode to show that as CPU Usage but it's ok. Now show us profiling data from your game please.

**Comment by Joncom** [comment]
Perhaps profiling data for a custom bunnymark build will suffice instead?

![screen shot 2019-02-12 at 6 26 02 pm](https://user-images.githubusercontent.com/1667179/52703976-23a1c800-2fb2-11e9-9488-aed554c9248e.png)

**Comment by RobDangerous** [comment]
Nope.

**Comment by Joncom** [comment]
Can we please stick to bunnymark? Because I'm having performance issues in bunnymark too, and it seems like a simpler test case to work with.

**Comment by RobDangerous** [comment]
What kind of performance problems do you have with the bunnymark?

**Comment by Joncom** [comment]
I wanted to benchmark the iPhone in a "worst case" scenario, so I [tweaked](https://github.com/Joncom/frameworks_test/tree/NO_BATCH/bunnymark/kha) bunnymark to use a different texture for each bunny, and thereby bypass batching.

0 bunnies = 60 FPS
30 bunnies = 55 FPS
40 bunnies = 50 FPS
50 bunnies = 36 FPS
60 bunnies = 30 FPS

Didn't think 30 bunnies was very many, but already 60 FPS is not possible.

**Comment by RobDangerous** [comment]
Yes, doesn't look good. Same data for Metal, OpenGL and Safari please (the bunny-fps numbers and the profiling data).

**Comment by Joncom** [comment]
**Metal:**

0 bunnies = 60 FPS
30 bunnies = 55 FPS
60 bunnies = 30 FPS

![screen shot 2019-02-12 at 6 26 02 pm](https://user-images.githubusercontent.com/1667179/52703976-23a1c800-2fb2-11e9-9488-aed554c9248e.png)

**OpenGL:**

0 bunnies = 60 FPS
30 bunnies = 25 FPS
60 bunnies = 10 FPS

![screen shot 2019-02-13 at 6 11 34 pm](https://user-images.githubusercontent.com/1667179/52707707-eb52b780-2fba-11e9-933e-43d2c36eb3a2.png)

**Safari:**

0 bunnies = 60 FPS
30 bunnies = 57 FPS
60 bunnies = 31 FPS

Safari seems to be on par with Metal.

**Comment by RobDangerous** [comment]
Thanks, that's useful data. Strange though as the OpenGL code is the same on all platforms and tends to beat the competition. Will spend some time with my iPad during the weekend.

**Comment by RblSb** [comment]
@Joncom do you tested custom bunnymark bench with other platforms, like android-native/macos?

**Comment by Joncom** [comment]
@RblSb

**android-native (Samsung Galaxy S6):**

0 bunnies = 60 FPS
30 bunnies = 60 FPS
60 bunnies = 60 FPS
150  bunnies = 60 FPS
170 bunnies = 55 FPS
200 bunnies = 40 FPS

**OSX:**

Tried running on OSX, but got a segmentation fault error for some reason.

```text
$ /Users/Joncom/Desktop/frameworks_test/bunnymark/kha/build/osx-build/build/Release/BunnyMark.app/Contents/MacOS/BunnyMark
Initializing Haxe libraries
Starting Kore
Segmentation fault: 11
```

**Comment by RobDangerous** [comment]
I think I found the cause and the fix is in https://github.com/Kode/Kha/commit/23029d14221f388c9b9e490831670aabd4098172 - still have to test on iOS and port to the different backends though.

**Comment by Joncom** [comment]
Would testing iOS be as simple as checking out that “js_compat” branch and doing an iOS kha-make?

**Comment by RobDangerous** [comment]
No, the js_compat branch is not fit for general consumption, sorry. Also need to adapt the code changes to Metal.

**Comment by Joncom** [comment]
@RobDangerous, is this the file that the Metal changes need to be adapted into?

https://github.com/Kode/Kore/blob/master/Backends/Graphics5/Metal/Sources/Kore/VertexBuffer5Impl.mm

**Comment by RobDangerous** [comment]
Yes, needs a second unlock implementation which takes a count parameter. Not sure I merged everything that's needed from the jscompat branch though. I can merge in the evening (sorry for things taking so long, there's still that crazy project which I have to finish ASAP).

**Comment by Joncom** [comment]
https://github.com/Kode/Kore/commit/cc386d384fc8f570650bc1536db2ac2c60cf5944

Does this mean the fix is now implemented on iOS?

**Comment by Joncom** [comment]
Hi @RobDangerous. Hope all is well. Was just wondering if you've had any time/luck with this one?

**Comment by RobDangerous** [comment]
Ah, yes. I merged the things and it turned out that helped with everything but with Metal on iOS. So ironically OpenGL on iOS might be faster now. So still more to do.
